### PR TITLE
Add virtio-fs support

### LINKED ISSUE #9: Add support for virtio-fs
In order to provide filesystem sharing available between host and guest, cloud-hypervisor needs to implement the `virtio-fs` device that will soon land in the virtio specification.

### PR DESCRIPTION
**Pull request by sboeuf**
In order to support filesystem sharing between host and guest, this pull request introduces the support for `vhost-user-fs` (also known as `virtio-fs`) to its device model.

Fixes #9

### CODE REVIEW

**Comment by sboeuf** [comment]
Depends on:
- https://github.com/rust-vmm/vm-memory/pull/20
- https://github.com/jiangliu/vhostuser_rs/pull/7

File: vm-virtio/src/fs.rs [hunk 1]
```diff
@@ -0,0 +1,409 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
```
Comment (rbradford) [inline vm-virtio/src/fs.rs]: Is this correct for this whole file?

File: vm-virtio/src/fs.rs [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+        let avail_features = master.get_features().map_err(Error::VhostUserGetFeatures)?;
+        // Create virtio device config space.
+        // First by adding the tag.
+        let mut config_space = tag.to_string().into_bytes();
+        config_space.resize(CONFIG_SPACE_SIZE, 0);
+        // And then by copying the number of queues.
+        let mut num_queues_slice: [u8; 4] = unsafe { transmute((req_num_queues as u32).to_be()) };
+        num_queues_slice.reverse();
+        config_space[CONFIG_SPACE_TAG_SIZE..CONFIG_SPACE_SIZE].copy_from_slice(&num_queues_slice);
+
```
Comment (rbradford) [inline vm-virtio/src/fs.rs]: Can replace all this by `let num_queues_slice = req_num_queues.to_le_bytes()`

File: vm-virtio/src/fs.rs (hunk 1 above)
Comment (sboeuf) [reply vm-virtio/src/fs.rs]: Not sure. Well I mean no other project has this file or this code. I started it from scratch but of course it is inspired from the same model that we followed from Firecracker/Crosvm codebase. Do you think I should include their copyright?

File: vm-virtio/src/fs.rs (hunk 2 above)
Comment (sboeuf) [reply vm-virtio/src/fs.rs]: Yes :)
Thanks for finding this `u32` function, this is better than this ugly code...

File: vm-virtio/src/fs.rs (hunk 1 above)
Comment (rbradford) [reply vm-virtio/src/fs.rs]: This is fine then.

**Comment by bjzhjing** [comment]
@sboeuf Why don't you work based on vhost crate directly? Some differences there. If this work will be sent to upstream finally, I think change is needed, right?

**Comment by sboeuf** [comment]
@bjzhjing yes I need to rebase on top of vhost crate. The reason why I did this first was because I had a POC based off of vhost_user_rs crate and it was simpler to port. Also another reason is that vhost crate is not ready yet.

**Comment by sboeuf** [comment]
PR updated:
- Added the support for the memory to be backed by a file from the command line `--memory "size=256,file=/dev/shm"`
- Ported the code to the rust-vmm vhost crate (I had to submit a PR to fix a few things)

**Comment by sboeuf** [comment]
Updated on latest https://github.com/rust-vmm/vhost/pull/2

**Comment by sboeuf** [comment]
@rbradford @sameo @chao-p 
I updated the PR based on the PR rust-vmm/vm-memory#29, and also I copied the code from the pending PR rust-vmm/vhost#2 so that we don't depend on the non-existing vhost crate.

**Comment by sameo** [comment]
@sboeuf Are we waiting for a rust-vmm vm-memory PR to be merged here?

**Comment by sboeuf** [comment]
@sameo Yes we're waiting on https://github.com/rust-vmm/vm-memory/pull/29. Maybe you can have a look at https://github.com/rust-vmm/vm-memory/pull/29 if you have some spare time.

**Comment by sboeuf** [comment]
@sameo @chao-p This PR is updated and ready now that rust-vmm/vm-memory#29 has been merged :)

File: vmm/src/config.rs [hunk 3]
```diff
@@ -195,14 +192,89 @@ impl<'a> NetConfig<'a> {
...
 pub struct VmConfig<'a> {
     pub cpus: CpusConfig,
     pub memory: MemoryConfig,
     pub kernel: KernelConfig<'a>,
     pub cmdline: CmdlineConfig,
     pub disks: Vec<DiskConfig<'a>>,
-    pub rng: RngConfig<'a>,
     pub net: Option<NetConfig<'a>>,
+    pub rng: RngConfig<'a>,
+    pub fs: Option<FsConfig<'a>>,
```
Comment (sameo) [inline vmm/src/config.rs]: I think it would make sense to support multiple shared directories, so we should have a vector of `FsConfig` here if possible.

**Comment by sboeuf** [comment]
@sameo yes let me address your comments and I'll repush asap.

**Comment by sboeuf** [comment]
@sameo PR updated!

**Pull request merged**